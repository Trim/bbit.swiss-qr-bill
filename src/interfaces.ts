export interface IQRBill {
  version?: QRBillVersion;
  amount: number;
  currency: QRBillCurrency;
  account: string;
  creditor: IQRBillAddress;
  reference: string;
  debtor: IQRBillAddress;
  unstructeredMessage?: string;
  billInformation?: string;
  language: QRBillLanguage;
}

export interface IQRBillAddress {
  type: QRBillAddressType;
  name: string;
  address?: string;
  street?: string;
  buildingNumber?: string;
  postalCode: string;
  locality: string;
  country: string;
}

export enum QRBillCurrency {
  CHF = 'CHF',
  EUR = 'EUR',
  USD = 'USD',
}

export enum QRBillLanguage {
  DE = 'de-CH',
  FR = 'fr-CH',
  IT = 'it-CH',
  EN = 'en-US',
}

export enum QRBillVersion {
  V2_0 = '0200',
}

export enum QRBillAddressType {
  UNSTRUCTURED = 'K',
  STRUCTURED = 'S',
}

export interface IQRBillTranslations {
  paymentPart: string;
  accountPayableTo: string;
  reference: string;
  additionalInfo: string;
  currency: string;
  amount: string;
  receipt: string;
  acceptancePoint: string;
  payableBy: string;
  payableByNameAddr: string;
}
