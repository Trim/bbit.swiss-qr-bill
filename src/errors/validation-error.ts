export class QRBillValidationError extends Error {
  private _errors: string[];

  public constructor(errors: string[]) {
    super('Error while validating parameters');
    this._errors = errors;
  }

  public getValidationErrors(): string[] {
    return this._errors;
  }
}
