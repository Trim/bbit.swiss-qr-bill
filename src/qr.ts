import { ReferenceValidator } from './reference/reference';
import { isNodeJs } from './utils';
import { swissCorssImage } from './swiss-cross';
import { QRData } from './qr-data';
import { QRBillValidationError } from './errors/validation-error';
import { IQRBill, QRBillVersion, IQRBillAddress } from './interfaces';
import { IBAN } from './iban/iban';
import * as qrcode from 'qrcode';
import { Image, Canvas } from 'canvas';

export class QRCodeGenerator {
  private _iban = new IBAN();
  private _reference = new ReferenceValidator();

  public async generate(params: IQRBill): Promise<Blob | Buffer> {
    const data = this.generateQRCodeContent(params);

    const canvas = await qrcode.toCanvas(this._createCanvas(), data, { margin: 0 });

    // adding logo at center
    const imgDim = { width: 30, height: 30 }; //logo dimention
    const context = canvas.getContext('2d');
    const imageObj = this._createImage();
    imageObj.src = swissCorssImage;
    context.drawImage(imageObj, canvas.width / 2 - imgDim.width / 2, canvas.height / 2 - imgDim.height / 2, imgDim.width, imgDim.height);

    if (isNodeJs) {
      return canvas.toBuffer();
    } else {
      /* istanbul ignore next: not tesed with jest */
      return canvas.toBlob();
    }
  }

  public generateQRCodeContent(params: IQRBill): string {
    this._setDefaultVersionIfMissing(params);
    this._verifyParams(params);
    switch (params.version) {
      case QRBillVersion.V2_0:
        const data = new QRData();
        data.add('SPC');
        data.add('0200');
        data.add('1');
        data.add(this._iban.electronicFormat(params.account));
        data.add('K');
        data.add(params.creditor.name);
        data.add(params.creditor.address);
        data.add(params.creditor.postalCode + ' ' + params.creditor.locality);
        data.add();
        data.add();
        data.add(params.creditor.country);
        data.add();
        data.add();
        data.add();
        data.add();
        data.add();
        data.add();
        data.add();
        data.add(this._parseAmount(params.amount));
        data.add(params.currency);
        data.add('K');
        data.add(params.debtor.name);
        data.add(params.debtor.address);
        data.add(params.debtor.postalCode + ' ' + params.debtor.locality);
        data.add();
        data.add();
        data.add(params.creditor.country);
        data.add(this._iban.isQRIBAN(params.account) ? 'QRR' : 'SCOR');
        data.add(params.reference);
        data.add(params.unstructeredMessage);
        data.add('EPD');
        data.add();
        return data.toString();
      default:
        throw new Error(`QR bill version ${params.version} is not supported`);
    }
  }

  private _verifyParams(params: IQRBill): void {
    const errors: string[] = [];

    if (!params.account) {
      errors.push("Property 'account' has to be defined");
    } else {
      if (!this._iban.isValid(params.account)) {
        errors.push(`Property 'account' (IBAN) ${params.account} is not valid`);
      }
    }

    if (!params.reference) {
      errors.push("Property 'reference' has to be defined");
    } else {
      if (this._iban.isQRIBAN(params.account)) {
        if (!this._reference.isQRReference(params.reference) || !this._reference.isQRReferenceValid(params.reference)) {
          errors.push("Property 'reference' is not valid");
        }
      } else {
        if (this._reference.isQRReference(params.reference) || !this._reference.isReferenceValid(params.reference)) {
          errors.push("Property 'reference' is not valid");
        }
      }
    }

    if (!params.currency) {
      errors.push("Property 'currency' has to be defined");
    }

    if (!params.amount) {
      errors.push("Property 'amount' has to be defined");
    }

    errors.push(...this._verifyAddress(params.creditor, 'creditor'));
    errors.push(...this._verifyAddress(params.debtor, 'debtor'));

    if (errors.length > 0) {
      throw new QRBillValidationError(errors);
    }
  }

  private _verifyAddress(address: IQRBillAddress, type: 'creditor' | 'debtor'): string[] {
    const errors: string[] = [];

    if (!address) {
      errors.push(`Property '${type}' has to be defined`);
      return errors;
    }

    if (!address.name) {
      errors.push(`Property 'name' on '${type}' has to be defined`);
    }

    if (!address.address) {
      errors.push(`Property 'address' on '${type}' has to be defined`);
    }

    if (!address.postalCode) {
      errors.push(`Property 'postalCode' on '${type}' has to be defined`);
    }

    if (!address.locality) {
      errors.push(`Property 'locality' on '${type}' has to be defined`);
    }

    if (!address.country) {
      errors.push(`Property 'country' on '${type}' has to be defined`);
    }

    return errors;
  }

  private _setDefaultVersionIfMissing(params: IQRBill): void {
    if (!params.version) {
      params.version = QRBillVersion.V2_0;
    }
  }

  private _parseAmount(amount: number): string {
    return Number(amount).toFixed(2);
  }

  private _createCanvas(): Canvas | HTMLCanvasElement {
    if (isNodeJs) {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const Canvas = require('canvas');
      return Canvas.createCanvas(500, 500);
    } else {
      /* istanbul ignore next: not tesed with jest */
      return document.createElement('canvas');
    }
  }

  private _createImage(): Image | HTMLImageElement {
    if (isNodeJs) {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const Canvas = require('canvas');
      return new Canvas.Image();
    } else {
      /* istanbul ignore next: not tesed with jest */
      return document.createElement('img');
    }
  }
}
