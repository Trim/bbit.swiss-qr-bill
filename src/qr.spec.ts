import { QRBillValidationError } from './errors/validation-error';
import { IQRBill } from './interfaces';
import { QRCodeGenerator } from './qr';
import * as _ from 'lodash';

const qr = new QRCodeGenerator();

const defaultData: IQRBill = {
  account: 'CH2830000011623852950',
  amount: 100.0,
  currency: 'CHF',
  creditor: {
    name: 'bbit gmbh',
    address: 'Rainweg 10',
    postalCode: '3612',
    locality: 'Steffisburg',
    country: 'CH',
  },
  reference: '000000000000000012312312316',
  debtor: {
    name: 'Test AG',
    address: 'Musterstrasse 1',
    postalCode: '3600',
    locality: 'Thun',
    country: 'CH',
  },
  unstructeredMessage: 'Test message',
  billInformation: 'Test billing information',
};

describe('QR test', () => {
  it('Usnupported version should fail', () => {
    let error: QRBillValidationError;
    const cloned = _.cloneDeep(defaultData);
    (cloned as any).version = 'unsupported';
    try {
      qr.generateQRCodeContent(cloned);
    } catch (err) {
      error = err;
    }
    expect(error).toEqual(new Error(`QR bill version ${cloned.version} is not supported`));
  });

  it('Missing account (IBAN) should fail', () => {
    let error: QRBillValidationError;
    const cloned = _.cloneDeep(defaultData);
    delete cloned.account;
    try {
      qr.generateQRCodeContent(cloned);
    } catch (err) {
      error = err;
    }
    expect(error).toBeDefined();
    expect(error.getValidationErrors()).toHaveLength(2);
    expect(error.getValidationErrors()[0]).toBe("Property 'account' has to be defined");
  });

  it('Invalid account (IBAN) should fail', () => {
    let error: QRBillValidationError;
    const cloned = _.cloneDeep(defaultData);
    cloned.account = 'CH2830000011623852951';
    try {
      qr.generateQRCodeContent(cloned);
    } catch (err) {
      error = err;
    }
    expect(error).toBeDefined();
    expect(error.getValidationErrors()).toHaveLength(1);
    expect(error.getValidationErrors()[0]).toBe(`Property 'account' (IBAN) ${cloned.account} is not valid`);
  });

  it('Missing creditor should fail', () => {
    let error: QRBillValidationError;
    const cloned = _.cloneDeep(defaultData);
    delete cloned.creditor;
    try {
      qr.generateQRCodeContent(cloned);
    } catch (err) {
      error = err;
    }
    expect(error).toBeDefined();
    expect(error.getValidationErrors()).toHaveLength(1);
    expect(error.getValidationErrors()[0]).toBe("Property 'creditor' has to be defined");
  });

  it('Missing creditor properties should fail', () => {
    let error: QRBillValidationError;
    const cloned: any = _.cloneDeep(defaultData);
    cloned.creditor = {};
    try {
      qr.generateQRCodeContent(cloned);
    } catch (err) {
      error = err;
    }
    expect(error).toBeDefined();
    expect(error.getValidationErrors()).toHaveLength(5);
    expect(error.getValidationErrors()[0]).toBe("Property 'name' on 'creditor' has to be defined");
    expect(error.getValidationErrors()[1]).toBe("Property 'address' on 'creditor' has to be defined");
    expect(error.getValidationErrors()[2]).toBe("Property 'postalCode' on 'creditor' has to be defined");
    expect(error.getValidationErrors()[3]).toBe("Property 'locality' on 'creditor' has to be defined");
    expect(error.getValidationErrors()[4]).toBe("Property 'country' on 'creditor' has to be defined");
  });

  it('Missing debtor should fail', () => {
    let error: QRBillValidationError;
    const cloned = _.cloneDeep(defaultData);
    delete cloned.debtor;
    try {
      qr.generateQRCodeContent(cloned);
    } catch (err) {
      error = err;
    }
    expect(error).toBeDefined();
    expect(error.getValidationErrors()).toHaveLength(1);
    expect(error.getValidationErrors()[0]).toBe("Property 'debtor' has to be defined");
  });

  it('Missing debtor properties should fail', () => {
    let error: QRBillValidationError;
    const cloned: any = _.cloneDeep(defaultData);
    cloned.debtor = {};
    try {
      qr.generateQRCodeContent(cloned);
    } catch (err) {
      error = err;
    }
    expect(error).toBeDefined();
    expect(error.getValidationErrors()).toHaveLength(5);
    expect(error.getValidationErrors()[0]).toBe("Property 'name' on 'debtor' has to be defined");
    expect(error.getValidationErrors()[1]).toBe("Property 'address' on 'debtor' has to be defined");
    expect(error.getValidationErrors()[2]).toBe("Property 'postalCode' on 'debtor' has to be defined");
    expect(error.getValidationErrors()[3]).toBe("Property 'locality' on 'debtor' has to be defined");
    expect(error.getValidationErrors()[4]).toBe("Property 'country' on 'debtor' has to be defined");
  });

  it('Missing reference should fail', () => {
    let error: QRBillValidationError;
    const cloned = _.cloneDeep(defaultData);
    delete cloned.reference;
    try {
      qr.generateQRCodeContent(cloned);
    } catch (err) {
      error = err;
    }
    expect(error).toBeDefined();
    expect(error.getValidationErrors()).toHaveLength(1);
    expect(error.getValidationErrors()[0]).toBe("Property 'reference' has to be defined");
  });

  it('Wrong reference should fail', () => {
    let error: QRBillValidationError;
    const cloned = _.cloneDeep(defaultData);
    cloned.reference = '1234';
    try {
      qr.generateQRCodeContent(cloned);
    } catch (err) {
      error = err;
    }
    expect(error).toBeDefined();
    expect(error.getValidationErrors()).toHaveLength(1);
    expect(error.getValidationErrors()[0]).toBe("Property 'reference' is not valid");
  });

  it('Missing currency should fail', () => {
    let error: QRBillValidationError;
    const cloned = _.cloneDeep(defaultData);
    delete cloned.currency;
    try {
      qr.generateQRCodeContent(cloned);
    } catch (err) {
      error = err;
    }
    expect(error).toBeDefined();
    expect(error.getValidationErrors()).toHaveLength(1);
    expect(error.getValidationErrors()[0]).toBe("Property 'currency' has to be defined");
  });

  it('Missing amount should fail', () => {
    let error: QRBillValidationError;
    const cloned = _.cloneDeep(defaultData);
    delete cloned.amount;
    try {
      qr.generateQRCodeContent(cloned);
    } catch (err) {
      error = err;
    }
    expect(error).toBeDefined();
    expect(error.getValidationErrors()).toHaveLength(1);
    expect(error.getValidationErrors()[0]).toBe("Property 'amount' has to be defined");
  });

  it('generateQRCodeContent should work', () => {
    const data = qr.generateQRCodeContent(defaultData);
    expect(data).toBeDefined();
    expect(data).toMatchSnapshot();
  });

  it('generate should work', async () => {
    const data = await qr.generate(defaultData);
    expect(data).toBeDefined();
    expect(data).toMatchSnapshot();
  });

  // it('generate should work', () => {
  //   const code = qr.generate(defaultData);
  // });
});
