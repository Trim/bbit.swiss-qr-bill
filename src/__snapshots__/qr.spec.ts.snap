// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QR test generateQRCodeContent should work 1`] = `
"SPC
0200
1
CH9300762011623852957
K
bbit gmbh
Rainweg 10
3612 Steffisburg


CH







100.00
CHF
K
Test AG
Musterstrasse 1
3600 Thun


CH
SCOR
1234
Test message
EPD

"
`;
